core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -70,21 +77,34 @@ class LogSegment(val log: FileMessageSet,
    *
    * It is assumed this method is being called from within a lock.
    *
-   * @param offset The first offset in the message set.
+   * @param firstOffset The first offset in the message set.
+   * @param largestTimestamp The largest timestamp in the message set.
    * @param messages The messages to append.
    */
   @nonthreadsafe
-  def append(offset: Long, messages: ByteBufferMessageSet) {
+  def append(firstOffset: Long, largestTimestamp: Long, messages: ByteBufferMessageSet) {
     if (messages.sizeInBytes > 0) {
-      trace("Inserting %d bytes at offset %d at position %d".format(messages.sizeInBytes, offset, log.sizeInBytes()))
+      trace("Inserting %d bytes at offset %d at position %d".format(messages.sizeInBytes, firstOffset, log.sizeInBytes()))
+      val physicalPosition = log.sizeInBytes()
+      val maxTimestampSoFarBeforeAppend = maxTimestampSoFar
+      // append the messages
+      log.append(messages)
+      // Update the in memory max timestamp and corresponding offset.
+      if (largestTimestamp > maxTimestampSoFar) {
+        // Always append the timeindex entry when we see the first message with a time index.
+        // The first time index entry is used by time based log rolling.
+        if (maxTimestampSoFar == Message.NoTimestamp)
+          timeIndex.maybeAppend(largestTimestamp, firstOffset)
```

Comment:
Yes, I was thinking of just looking at the timestamp of the first message. If it doesn't have timestamp, we use create time. This is probably simpler than having to force an index entry on first message with timestamp. For those segments with a mix of valid timestamp and NoTimestamp, we can decide whether the rolling will be based on the old or the new behavior. Either is fine, but we probably want to pick one that's easier to implement.